### PR TITLE
Gateway Payloads

### PR DESCRIPTION
**Pull request by ianagbip1oti**
Should be reviewed after #29 

Adds gateway payloads
Refactors some of the codegen code to keep it DRY

### CODE REVIEW

File: src/codegen/ruby/def/event.rb [hunk 1]
```diff
@@ -1,52 +0,0 @@
-
-require_relative 'field'
-
-class EventDef
```
Comment (CoreyShupe) [inline src/codegen/ruby/def/event.rb]: Why are we removing this?

File: src/codegen/resources/gateway/connection_properties [hunk 2]
```diff
@@ -0,0 +1,3 @@
+$os	string	your operating system
```
Comment (CoreyShupe) [inline src/codegen/resources/gateway/connection_properties]: This is pretty inconsistent with how we did Routes and their Data classes. Should we consider moving all of their data classes here and only having the routes in .Java.

File: src/codegen/ruby/def/resource.rb [hunk 3]
```diff
@@ -2,11 +2,12 @@
 require_relative 'field'
 
 class ResourceDef
-  attr_reader :name, :fields
+  attr_reader :name, :fields, :imports
   
-  def initialize(name, fields)
-    @name = name
+  def initialize(name, fields, opts={})
```
Comment (CoreyShupe) [inline src/codegen/ruby/def/resource.rb]: What is the opts={} for? It seems like a pretty random addition, just not sure what it does.

File: src/codegen/ruby/java_types.rb [hunk 4]
```diff
@@ -26,12 +26,14 @@
   'array of snowflakes' => 'ImmutableList<Snowflake>',
   'array of strings' => 'ImmutableList<String>',
   'array of two integers (current_size, max_size)' => 'ImmutableList<Long>',
+  'array of two integers (shard_id, num_shards)' => 'ImmutableList<Long>',
```
Comment (CoreyShupe) [inline src/codegen/ruby/java_types.rb]: Is an ImmutableList a good idea here? Or should we instead maybe move the Shard class to Data.

File: src/codegen/ruby/def/event.rb (hunk 1 above)
Comment (ianagbip1oti) [reply src/codegen/ruby/def/event.rb]: Instead the ResourceDef class is used. They were very close to identical, with the minor differences now set by configuration.

File: src/codegen/resources/gateway/connection_properties (hunk 2 above)
Comment (ianagbip1oti) [reply src/codegen/resources/gateway/connection_properties]: I believe with the routes the inputs and such were not as consistent with how they are formatted in the discord docs.

If we can move them here as codegen, then we should.

File: src/codegen/ruby/def/resource.rb (hunk 3 above)
Comment (ianagbip1oti) [reply src/codegen/ruby/def/resource.rb]: https://stackoverflow.com/a/18407748

File: src/codegen/ruby/java_types.rb (hunk 4 above)
Comment (ianagbip1oti) [reply src/codegen/ruby/java_types.rb]: ImmutableList is consistent with the other array of two integers there. We can still use the Shard class and convert to ImmutableList<Integer> before it's sent.

File: src/codegen/resources/gateway/connection_properties (hunk 2 above)
Comment (CoreyShupe) [reply src/codegen/resources/gateway/connection_properties]: Then we should, maybe in another PR for Data once the rest of codegen works. Less priority

File: src/codegen/ruby/java_types.rb (hunk 4 above)
Comment (CoreyShupe) [reply src/codegen/ruby/java_types.rb]: Where in the discord docs does it use that terminology. I find it very dangerous considering Lists aren't meant to just contain 2 values or really any set number. They're more defined as a list of things when here that really isnt the case, it's a pair of items.

Comment (ianagbip1oti) [reply src/codegen/ruby/java_types.rb]: Everywhere we have arrays, we use an ImmutableList. Raw arrays don't work as well with modern java (e.g., with generics and streams).

It's true that we could use a Pair or some other type, but it's going to be appear in JSON as any other array does.

Comment (CoreyShupe) [reply src/codegen/ruby/java_types.rb]: We shouldn't fudge it, we can make a custom deserializer for the Shard and have the Shard class move over to Data. It really isnt an array either it's just displayed that way.

Comment (ianagbip1oti) [reply src/codegen/ruby/java_types.rb]: Added classes to handle these

**Comment by CoreyShupe** [comment]
This should be merged, we should think about moving Route resources into here though to maintain consistency.

**Pull request merged**